### ISSUE TITLE
FullHttpRequestMatcher: not identical data returns success response

### ISSUE THREAD

**Issue by katjai1** [lomsa-dev/http-mock-adapter#149]
Not identical data returns success response, probably need to add strict data check to the FullHttpRequestMatcher

```
    test('Not identical data and FullHttpRequestMatcher', () async {
      var dio = Dio();
      var dioAdapter = DioAdapter(dio: dio);
      var path = '/abc';
      var data = {'a': 1, 'b': 2, 'c': 3, 'd': 4, 'e': 5};
      var dataExpected = {'a': 1, 'b': 2};
      dioAdapter.onPost(
        path,
        (server) => server.reply(
          200,
          'OK',
        ),
        data: dataExpected,
      );

      expect(
        () async => await dio.post(path, data: data),
        throwsA(isA<DioException>()),
      );
    });

HTTP mock test test Not identical data and FullHttpRequestMatcher [E]                                                                                                                                                                                        
  Expected: throws <Instance of 'DioException'>
    Actual: <Closure: () => Future<Response<dynamic>>>
     Which: returned a Future that emitted Response<dynamic>:<OK>
```

**Comment by sebastianbuechler** [comment]
It seems that currently the request body check is not strict as this test on the `matches_request_test.dart` will fail:
```dart

        test('actual is a subset of expected', () {
          const actual = {
            'a': 'a',
            'b': 'b',
            'c': 'c',
          };

          const expected = {
            'a': 'a',
            'b': 'b',
          };

          expect(options.matches(actual, expected), false);
        });
```
@LukaGiorgadze Is this behavior intended? 

We could reason that as long as the required subset is present in the body the condition is fulfilled, as usually that's how most web servers will also evaluate if the request body is good or bad.

Maybe there is demand for a new request matcher that only allows strict body checks?

**Comment by LukaGiorgadze** [comment]
@sebastianbuechler Yeh, that makes sense.
Adding a new checker that checks nested data like deep equal is worth it.

Thanks @katjai1

**Comment by sebastianbuechler** [comment]
This is solved in #158 and will be available in the next release.

Usage: 
```dart
      dio = Dio(BaseOptions(contentType: Headers.jsonContentType));
      dioAdapter = DioAdapter(
        dio: dio,
        matcher: const FullHttpRequestMatcher(needsExactBody: true),
      );
```

`needsExactBody` is defaulting to `false`, so no change is needed for existing code.
